Fix for cron service not running by default in sm distribution 1.6 for JupyterLab

lavaraja opened the pull request:
*Issue #, if available:*  https://github.com/aws-samples/sagemaker-studio-apps-lifecycle-config-examples/issues/12 

*Description of changes:*

1. Added fix for handling the scenario where cron service is not started by default in [SM Distribution 1.6.0](https://github.com/aws/sagemaker-distribution/issues/354)
2. Added logic for backward compatibility with SM Distribution 1.5.0 and below.

*Test results:*

Tested with JupterLab image SM Distribution 1.5.0 and SM Distribution 1.6.0. 

<img width="1208" alt="Screenshot 2024-04-21 at 19 28 49" src="https://github.com/aws-samples/sagemaker-studio-apps-lifecycle-config-examples/assets/3517075/4cafaced-89ec-44c6-9597-91f9b65495a7">
<img width="1134" alt="Screenshot 2024-04-21 at 19 28 40" src="https://github.com/aws-samples/sagemaker-studio-apps-lifecycle-config-examples/assets/3517075/515d1f21-a451-4666-803f-bec7032739e2">
<img width="1198" alt="Screenshot 2024-04-21 at 19 28 56" src="https://github.com/aws-samples/sagemaker-studio-apps-lifecycle-config-examples/assets/3517075/58142a4d-790a-4366-ba86-449e6538f1a4">
<img width="1182" alt="Screenshot 2024-04-21 at 19 28 19" src="https://github.com/aws-samples/sagemaker-studio-apps-lifecycle-config-examples/assets/3517075/9b554c8c-9d54-4177-bb1b-76e33ab57c06">
<img width="1180" alt="Screenshot 2024-04-21 at 19 28 27" src="https://github.com/aws-samples/sagemaker-studio-apps-lifecycle-config-examples/assets/3517075/fd3fbae1-1d74-42d2-84f7-a778588b2fc2">
<img width="1206" alt="Screenshot 2024-04-21 at 19 32 50" src="https://github.com/aws-samples/sagemaker-studio-apps-lifecycle-config-examples/assets/3517075/a7131338-c4f4-45c9-8ac2-120861b68540">


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

lavaraja committed:
Fix for verifying the cron package and cron service status

lavaraja committed:
Modified and added logic to handle non zero return code

lavaraja committed:
Fix for verifying and starting cron service.

edwardps commented on code-editor/auto-stop-idle/on-start.sh:
```diff
@@ -37,4 +54,4 @@ sudo /bin/bash -c "echo 'AWS_CONTAINER_CREDENTIALS_RELATIVE_URI=$AWS_CONTAINER_C
 # Add script to crontab for root.
 echo "Adding autostop idle Python script to crontab..."
-echo "*/2 * * * * /bin/bash -ic '$CONDA_HOME/python $PYTHON_SCRIPT_PATH --time $IDLE_TIME_IN_SECONDS --region $AWS_DEFAULT_REGION >> $LOG_FILE'" | sudo crontab -
+echo "*/2 * * * * /bin/bash -ic '$CONDA_HOME/python $PYTHON_SCRIPT_PATH --time $IDLE_TIME_IN_SECONDS --region $AWS_DEFAULT_REGION >> $LOG_FILE'" | sudo crontab -
```
Is this line actually changed?

aws-spenceng replied:
Were lines 40-56 removed?

lavaraja replied:
@edwardps no the below line was not changed. Line number 40 was moved to 57. 

`echo "*/2 * * * * /bin/bash -ic '$CONDA_HOME/python $PYTHON_SCRIPT_PATH --time $IDLE_TIME_IN_SECONDS --region $AWS_DEFAULT_REGION >> $LOG_FILE'" | sudo crontab -
`

@aws-spenceng nope. No lines were removed. We modified the logic in the line [20-21 ](https://github.com/aws-samples/sagemaker-studio-apps-lifecycle-config-examples/blob/9c305c0ef6fb2b71c7d9329e193ba719e9db9c1b/code-editor/auto-stop-idle/on-start.sh#L20) and added extra lines [21-38](https://github.com/lavaraja/sagemaker-studio-apps-lifecycle-config-examples/blame/112c43e36633b94d8ec2a0327881ba188e68fe77/code-editor/auto-stop-idle/on-start.sh#L21)

edwardps commented on code-editor/auto-stop-idle/on-start.sh:
```diff
@@ -17,8 +17,25 @@ PYTHON_SCRIPT_PATH=$SOLUTION_DIR/sagemaker_code_editor_auto_shut_down/auto_stop_
+	sudo apt install cron
+else
+	echo "Package cron is already installed."
+    # start/restart the service.
```
nit: indent.

lavaraja replied:
Corrected this. :)

edwardps commented on code-editor/auto-stop-idle/on-start.sh:
```diff
@@ -17,8 +17,25 @@ PYTHON_SCRIPT_PATH=$SOLUTION_DIR/sagemaker_code_editor_auto_shut_down/auto_stop_
+
+	# Installing cron.
+	echo "Installing cron..."
+	sudo apt install cron
```
Seems this is not working if it's running in a private vpc(no internet access). Can we add mention this in readme file? 
This is not a new issue from this PR.

lavaraja replied:
Added [steps](https://github.com/lavaraja/sagemaker-studio-apps-lifecycle-config-examples/tree/main/code-editor/auto-stop-idle#steps-for-installing-in-internet-free-vpc-environments) for installing in a private vpc(no internet access) and also additional steps required for sm distribution image version 1.5 and below(where cron package was not part of the image).

edwardps reviewed (commented):
Thanks for the contribution. Posted some comments.

hugowong-aws commented:
Hi, is adding the following command 
```
sudo service cron start
``` 
after installing cron
```
sudo apt-get install -y cron
```
not sufficient in unblocking the extension? I was able to get this to work on a SageMaker Studio JupyterLab app using SageMaker Distribution 1.7.

edwardps commented:
> Hi, is adding the following command
> 
> ```
> sudo service cron start
> ```
> 
> after installing cron
> 
> ```
> sudo apt-get install -y cron
> ```
> 
> not sufficient in unblocking the extension? I was able to get this to work on a SageMaker Studio JupyterLab app using SageMaker Distribution 1.7.

A suggestion is to move "sudo service cron restart" out the else block. (starting the cron can be shared for both case)

lavaraja committed:
Indentation adjusted.

lavaraja committed:
indentation adjusted

lavaraja committed:
Added steps for installing in Internet-Free VPC environments

lavaraja committed:
Updated steps for Steps for installing in Internet-Free VPC environments

lavaraja committed:
Updated Steps for installing in Internet-Free VPC environments:

lavaraja committed:
Update steps for installing in Internet-Free VPC environments

lavaraja committed:
Update Steps for installing in Internet-Free VPC environments.

lavaraja committed:
Update steps for Steps for installing in Internet-Free VPC environments

lavaraja committed:
Updated steps for installing in Internet-Free VPC environments

lavaraja committed:
Updated README.md

Line 45 will be moved to 56 after the PR. Updated it in the readme file.

Line number 46 will be moved to 56

curl -LO --output-dir /var/tmp/ https://github.com/aws-samples/sagemaker-studio-apps-lifecycle-config-examples/releases/download/v$ASI_VERSION/$PYTHON_PACKAGE

edwardps commented on code-editor/auto-stop-idle/on-start.sh:
```diff
@@ -17,8 +17,25 @@ PYTHON_SCRIPT_PATH=$SOLUTION_DIR/sagemaker_code_editor_auto_shut_down/auto_stop_
+else
+	echo "Package cron is already installed."
+        # Start/restart the service.
+	sudo service cron restart
```
Does installation auto enable and start the cron service?
Do need to start cron in the installation branch? (this is discussed above)

lavaraja replied:
Yes, the installation auto starts the cron service. Reference for the same. ↓

```
sagemaker-user@default:~$ sudo service cron status
cron: unrecognized service
sagemaker-user@default:~$ sudo sh -c 'printf "#!/bin/sh\nexit 0" > /usr/sbin/policy-rc.d'
sagemaker-user@default:~$ sudo apt install -y cron -q

invoke-rc.d: could not determine current runlevel
 * Starting periodic command scheduler cron                                                                                                       [ OK ] 
sagemaker-user@default:~$ sudo service cron status
 cron is running
sagemaker-user@default:~$ 
```
Do need to start cron in the installation branch? - no.

edwardps commented on jupyterlab/auto-stop-idle/on-start.sh:
```diff
@@ -29,13 +29,24 @@ STATE_FILE=$SOLUTION_DIR/auto_stop_idle.st
+	sudo apt install cron
+else
+	echo "Package cron is already installed."
+    # start/restart the service.
```
Nit: indent.

lavaraja replied:
Done.

edwardps commented on code-editor/auto-stop-idle/README.md:
````diff
@@ -50,6 +50,45 @@ aws sagemaker update-domain \
+
+1. Downloading autostop idle Python package and upload to s3 bucket.
+```
+ASI_VERSION=0.3.0
````
Should this be placeholder for customer to replace with the latest version?

lavaraja replied:
Yes.

edwardps commented on code-editor/auto-stop-idle/README.md:
````diff
@@ -50,6 +50,45 @@ aws sagemaker update-domain \
     }'
 
 ```
+## Steps for installing in Internet-Free VPC environments:
````
The private VPC takes more complexity. 
I would suggest to separate this support. 
Step #1 - address the cron issue in current PR. Leave a comment in readme. In doing this, we can quickly close the current PR which already adds value for customers. 
Step #2 - follow up on PR #2 to provide private VPC support. Currently, in the readme providing manual steps provide sub-optimal user experience. We need more time to simplify the UX of private VPC case.

lavaraja replied:
Removed steps for private VPC support and added a note in the readme file. As suggested will work on this as a separate PR after simplifying the UX of private VPC case.

lavaraja committed:
indented.

Adjusted the indentation.

edwardps commented on code-editor/auto-stop-idle/README.md:
````diff
@@ -50,6 +50,45 @@ aws sagemaker update-domain \
+aws s3 cp sagemaker_code_editor_auto_shut_down-$ASI_VERSION.tar.gz <YOUR-S3-BUCKET-PATH> 
+```
+
+2. Edit the on-start.sh file and replace "45 and 46" with:
````
> replace "45 and 46" 
This is easy to be obsolete after we further change the script and it would be hard to remember to update this.

lavaraja committed:
Removed the installation steps for Internet free VPC free environments.

1.Removed the installation steps for Internet free VPC free environments.

2. Added a note in readme file mentioning that currently it is not supported in internet free VPC environments.
